text/0032-minimal-relay.md

```diff
@@ -0,0 +1,190 @@
+# RFC-0032: Minimal Relay
+
+|                 |                                                                               |
+| --------------- | ----------------------------------------------------------------------------- |
+| **Start Date**  | 20 September 2023                                                             |
+| **Description** | Proposal to minimise Relay Chain functionality.                               |
+| **Authors**     | Joe Petrowski, Gavin Wood                                                     |
+
+## Summary
+
+The Relay Chain contains most of the core logic for the Polkadot network. While this was necessary
+prior to the launch of parachains and development of XCM, most of this logic can exist in
+parachains. This is a proposal to migrate several subsystems into system parachains.
+
+## Motivation
+
+Polkadot's scaling approach allows many distinct state machines (known generally as parachains) to
+operate with common guarantees about the validity and security of their state transitions. Polkadot
+provides these common guarantees by executing the state transitions on a strict subset (a backing
+group) of the Relay Chain's validator set.
+
+However, state transitions on the Relay Chain need to be executed by _all_ validators. If any of
+those state transitions can occur on parachains, then the resources of the complement of a single
+backing group could be used to offer more cores. As in, they could be offering more coretime (a.k.a.
+blockspace) to the network.
+
+By minimising state transition logic on the Relay Chain by migrating it into "system chains" -- a
+set of parachains that, with the Relay Chain, make up the Polkadot protocol -- the Polkadot
+Ubiquitous Computer can maximise its primary offering: secure blockspace.
+
+## Stakeholders
+
+- Parachains that interact with affected logic on the Relay Chain;
+- Core protocol and XCM format developers;
+- Tooling, block explorer, and UI developers.
+
+## Explanation
+
+The following pallets and subsystems are good candidates to migrate from the Relay Chain:
+
+- Identity
+- Preimage
```

Comment:
It seems hard to migrate the preimages from the relay chain since these are almost by definition storage items which are heavy and large. These need to stay on the relay chain so that there isnt some kind of PoV limit which bottlenecks access to preimages.

Reply:
Many of the medium term scaling solutions to me actually involve allowing others to post data onto the relay chain directly, and having mechanisms which allow the parachain validation to access those preimages on the relay chian.

https://github.com/paritytech/polkadot-sdk/issues/811

Reply:
It's only mentioned here, not elsewhere, so not sure what this means, but..

In principle wasm or risc-v preimage could live in availability, where they get reposted/retained periodically.  We want all artifacts to be prebuilt on all the validators, so the validators would've downloaded all the preimages in the past, built them, and retained the build artifacts.  Any new validators fetch them from availability.  I seriously the savings here warrant the development time, but who knows maybe this becomes easy.

Is more possible?  In principle artifacts could be prebuilt only by approval checkers, which proves the artifact to be buildable. We'd face several problems here:
- Artifact could build today but no longer build tomorrow, due to some parameter change, host upgrade, etc.
- Artifacts needed build times within our timeouts previously. It's possible our time overruns scheme makes this less problematic, not completely sure.  
- We've fairly slow wasm build times, but maybe risc-v improves this.
 
We'd always introduce unpredictability into resource usage this way, meaning this harms performance somewhat.  At the same time it's unclear what this really gains us.

Reply:
Isn't availability bandwidth the same limit which leads to a maximum 5 MB PoV?

Or what would be the expectations on nodes to have and retain data here?

The common use cases for on-relay-chain pre-images for the PoV to me are:

- large runtime upgrades
- commonly used smart contracts

Especially in the case of immutable smart contracts, seems very sensible to have them stored on on the relay chain directly.

Reply:
Fair enough, CoreJam also makes use of a Storage pallet that looks similar. I will remove this from the list of candidates.